### PR TITLE
Version mismatch playbook

### PR DESCRIPTION
**Pull request by Avi-Robusta**
Screen Shot 2024-02-21 at 16 22 17

### CODE REVIEW

**Comment by Avi-Robusta** [comment]
Screen Shot 2024-02-21 at 16 22 17

File: playbooks/robusta_playbooks/version_mismatch_enricher.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+            return False
+
+        # some prometheus's remove specific labels due to space constraints
+        if len([metric for metric in self.raw_metrics if metric.get("node") is not None]) == 0:
+            logging.error(f"Prometheus is missing the 'node' label, unable to detect node versions.")
+            return False
+
+        # you can have multiple versions of the api server in the metrics at once
+        # i.e. at the time of kubernetes upgrade both will show up in the metrics for several minutes
+        if len(self.api_versions) == 0:
```
Comment (arikalon1) [inline playbooks/robusta_playbooks/version_mismatch_enricher.py:48]: Can we make this more flexible, and support the case we don't have the api server version?
Like, if for example, we have 2 node pools, with different versions

File: playbooks/robusta_playbooks/version_mismatch_enricher.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    if query_result.result_type == "error" or query_result.vector_result is None:
+        logging.error(f"version_mismatch_enricher failed to get prometheus results.")
+        return
+
+    build_infos = BuildInfoResults(query_result)
+    if not build_infos.is_results_valid():
+        logging.error(f"Invalid prometheus results for version_mismatch_enricher.")
+        return
+
+    if len(build_infos.api_versions) > 1:
```
Comment (arikalon1) [inline playbooks/robusta_playbooks/version_mismatch_enricher.py:80]: I think I would also support the case we don't know the api server version (i.e. len(api_versions) == 0 )

File: playbooks/robusta_playbooks/version_mismatch_enricher.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+        api_version_string = f"has reported versions {', '.join(build_infos.api_versions)}"
+    else:
+        api_version_string = f"is version {build_infos.api_versions[0]}"
+
+    # in the case where a node is of a higher version than the api server
+    alert.add_enrichment(
+        [
+            MarkdownBlock(f"Automatic {alert.alert_name} investigation:"),
```
Comment (arikalon1) [inline playbooks/robusta_playbooks/version_mismatch_enricher.py:88]: I would remove this.
I think we don't need to state we're investigating, and just show the results

File: src/robusta/core/playbooks/prometheus_enrichment_utils.py [hunk 4]
```diff
@@ -350,9 +351,38 @@ def create_chart_from_prometheus_query(
...
+    For graphs use run_prometheus_query_range which uses the prometheus query_range api
+    """
+    try:
+        prom = get_prometheus_connect(prometheus_params)
+        query = __add_additional_labels(query, prometheus_params)
+        prom_params = {"timeout": PROMETHEUS_REQUEST_TIMEOUT_SECONDS}
+        prom.check_prometheus_connection(prom_params)
+        results = prom.safe_custom_query(query=query, params=prom_params)
+        return PrometheusQueryResult(results)
+    except Exception as e:
```
Comment (arikalon1) [inline src/robusta/core/playbooks/prometheus_enrichment_utils.py:377]: I think I would not surround with try/except
We have a global code to handle Prometheus errors

Also, someone calling the api might not expect having the error as a field of the result

**Review by arikalon1 — APPROVED** [review]
nice work
left few minor comments

**Comment by Avi-Robusta** [comment]
testing

**Pull request merged**